### PR TITLE
Mod encoder clicks show first current value, then the next click changes the value

### PR DESCRIPTION
**Pull request by soymonitus**

### CODE REVIEW

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 1]
```diff
@@ -156,11 +168,23 @@ void GlobalEffectable::displayCompressorAndReverbSettings(bool on) {
 	}
 	else {
 		if (on) {
-			display->displayPopup(getCompressorModeDisplayName());
+			if (getSelectedClip()) {
+				// We are inside a clip (either audioclip or affect-entire kit): show Sidechain sync
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:172]: The comment should just say "We are inside a clip : show Sidechain sync" - the master compressor is not in any of the clips, so it should be side chain everywhere outside song.

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 2]
```diff
@@ -156,11 +168,23 @@ void GlobalEffectable::displayCompressorAndReverbSettings(bool on) {
...
+				if (compressor.syncLevel == SYNC_LEVEL_32ND) {
+					display->displayPopup("SLOW");
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:174]: use this for the string: l10n::get(STRING_FOR_SLOW_SIDECHAIN_COMPRESSOR);

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 3]
```diff
@@ -156,11 +168,23 @@ void GlobalEffectable::displayCompressorAndReverbSettings(bool on) {
...
+				}
+				else {
+					display->displayPopup("FAST");
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:177]: use this for the string: l10n::get(STRING_FOR_FAST_SIDECHAIN_COMPRESSOR)

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 4]
```diff
@@ -156,11 +168,23 @@ void GlobalEffectable::displayCompressorAndReverbSettings(bool on) {
 	}
 	else {
 		if (on) {
-			display->displayPopup(getCompressorModeDisplayName());
+			if (getSelectedClip()) {
+				// We are inside a clip (either audioclip or affect-entire kit): show Sidechain sync
+				if (compressor.syncLevel == SYNC_LEVEL_32ND) {
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:173]: is this using the same logic as Sound::displaySidechainAndReverbSettings and Sound::getSidechainDisplayName?

**Review by seangoodvibes — CHANGES_REQUESTED** [review]

File: src/deluge/model/global_effectable/global_effectable.cpp (hunk 4 above)
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:173]: Just tested and this doesn't work. It's changing between Slow and Fast with every press. I think you need to the same logic as in Sound::

**Review by seangoodvibes — CHANGES_REQUESTED** [review]

File: src/deluge/model/global_effectable/global_effectable.cpp (hunk 4 above)
Comment (soymonitus) [reply src/deluge/model/global_effectable/global_effectable.cpp:173]: Yes, it should be

**Comment by soymonitus** [comment]
Ready to test and review

File: src/deluge/gui/views/view.cpp [hunk 5]
```diff
@@ -1176,6 +1176,10 @@ void View::modButtonAction(uint8_t whichButton, bool on) {
 		if (on) {
 
 			if (isUIModeWithinRange(modButtonUIModes) || (getRootUI() == &performanceSessionView)) {
+				// change the button selection before calling mod button action so that mod button action
+				// knows the mod button parameter context
+				*activeModControllableModelStack.modControllable->getModKnobMode() = whichButton;
+
 				activeModControllableModelStack.modControllable->modButtonAction(
 				    whichButton, true, (ParamManagerForTimeline*)activeModControllableModelStack.paramManager);
```
Comment (seangoodvibes) [inline src/deluge/gui/views/view.cpp:1185]: The line below (1186) is duplicated and needs to be removed as it was moved above (1181)

File: src/deluge/model/global_effectable/global_effectable_for_clip.cpp [hunk 6]
```diff
@@ -223,24 +224,78 @@ int32_t GlobalEffectableForClip::getParameterFromKnob(int32_t whichModEncoder) {
...
+
+void GlobalEffectableForClip::displaySidechainAndReverbSettings(bool on) {
+	if (display->haveOLED()) {
+		if (on) {
+			DEF_STACK_STRING_BUF(popupMsg, 100);
+			popupMsg.append("Sidechain: ");
+			popupMsg.append(getSidechainDisplayName());
+			popupMsg.append("\n");
+
+			// Revern
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable_for_clip.cpp:244]: Nit: Typo, "Reverb"

**Review by seangoodvibes — CHANGES_REQUESTED** [review]

File: src/deluge/gui/views/view.cpp [hunk 7]
```diff
@@ -1489,18 +1491,21 @@ void View::pretendModKnobsUntouchedForAWhile() {
...
 
 void View::cycleThroughReverbPresets() {
-
+	// Get the current value to a tmp variable
 	int32_t currentPreset = getCurrentReverbPreset();
-
-	int32_t newPreset = currentPreset + 1;
+	int32_t newPreset = currentPreset;
+	// Change the tmp value (only if popup is already showing)
+	newPreset = currentPreset + 1;
```
Comment (seangoodvibes) [inline src/deluge/gui/views/view.cpp:1498]: I think we can remove references to tmp value here, and just leave it as `int32_t newPreset = currentPreset + 1;`

File: src/deluge/gui/views/view.cpp [hunk 8]
```diff
@@ -1489,18 +1491,21 @@ void View::pretendModKnobsUntouchedForAWhile() {
...
 		newPreset = 0;
 	}
-
-	AudioEngine::reverb.setRoomSize((float)presetReverbRoomSize[newPreset] / 50);
-	AudioEngine::reverb.setDamping((float)presetReverbDampening[newPreset] / 50);
-
-	display->displayPopup(getReverbPresetDisplayName(newPreset));
+	// Show popup (may be the original value or the changed value)
+	display->popupTextTemporary(getReverbPresetDisplayName(newPreset), DisplayPopupType::MOD_ENCODER_CYCLE);
+	// Write the change
```
Comment (seangoodvibes) [inline src/deluge/gui/views/view.cpp:1504]: Are these changes leftover from the mod encoder popup change? I think you can put back what was there before.

File: src/deluge/hid/display/seven_segment.h [hunk 9]
```diff
@@ -62,10 +62,10 @@ class SevenSegment : public Display {
 
 	void consoleText(char const* text) override { SevenSegment::displayPopup(text); }
 	void popupText(char const* text, DisplayPopupType type = DisplayPopupType::GENERAL) override {
-		SevenSegment::displayPopup(text, type);
+		SevenSegment::displayPopup(text, 0, false, 255, 1, type);
 	}
 	void popupTextTemporary(char const* text, DisplayPopupType type = DisplayPopupType::GENERAL) override {
-		SevenSegment::displayPopup(text, type);
+		SevenSegment::displayPopup(text, 3, false, 255, 1, type);
 	}
```
Comment (seangoodvibes) [inline src/deluge/hid/display/seven_segment.h:69]: Will these changes affect anything or they are ok to stay?

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 10]
```diff
@@ -169,26 +172,31 @@ void GlobalEffectable::displayCompressorAndReverbSettings(bool on) {
...
-char const* GlobalEffectable::getCompressorModeDisplayName() {
-	return editingComp ? "FULL" : "ONE";
+char const* GlobalEffectable::getCompressorModeDisplayName(bool isEditingComp) {
+	return isEditingComp ? "FULL" : "ONE";
 }
 
-char const* GlobalEffectable::getCompressorParamDisplayName() {
-	currentCompParam = static_cast<CompParam>(util::to_underlying(currentCompParam) % maxCompParam);
+char const* GlobalEffectable::getCompressorParamDisplayName(CompParam compParam) {
+	CompParam tmpCompParam = static_cast<CompParam>(util::to_underlying(compParam) % maxCompParam);
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:180]: do we need this tmp stuff as we reverted the mod encoder change?

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 11]
```diff
@@ -268,10 +276,18 @@ bool GlobalEffectable::modEncoderButtonAction(uint8_t whichModEncoder, bool on,
 				                     == RuntimeFeatureStateToggle::Off)
 				                        ? (kNumModFXTypes - 1)
 				                        : kNumModFXTypes;
-				modFXType = static_cast<ModFXType>((util::to_underlying(modFXType) + 1) % modTypeCount);
-				if (modFXType == ModFXType::NONE) {
-					modFXType = static_cast<ModFXType>(1);
+				// Get the current value to a tmp variable
+				ModFXType tmpModFXType = modFXType;
+				// Change the tmp value
+				tmpModFXType = static_cast<ModFXType>((util::to_underlying(tmpModFXType) + 1) % modTypeCount);
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:282]: Do we need this tmp stuff? I think it's leftover from your mod encoder button changes.

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 12]
```diff
@@ -295,8 +311,7 @@ bool GlobalEffectable::modEncoderButtonAction(uint8_t whichModEncoder, bool on,
 				case ModFXType::NONE:
 					__builtin_unreachable();
 				}
-				display->displayPopup(displayText.data());
-				ensureModFXParamIsValid();
+				display->popupTextTemporary(displayText.data(), DisplayPopupType::MOD_ENCODER_CYCLE);
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:314]: I think this is leftover from mod encoder changes.

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 13]
```diff
@@ -305,10 +320,15 @@ bool GlobalEffectable::modEncoderButtonAction(uint8_t whichModEncoder, bool on,
 		}
 		else {
 			if (on) {
-				currentModFXParam =
-				    static_cast<ModFXParam>((util::to_underlying(currentModFXParam) + 1) % kNumModFXParams);
+				// Get the current value to a tmp variable
+				ModFXParam tmpModFXParam = currentModFXParam;
+				// Change the tmp value
+				tmpModFXParam = static_cast<ModFXParam>((util::to_underlying(tmpModFXParam) + 1) % kNumModFXParams);
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:326]: leftover from mod encoder changes

File: src/deluge/model/global_effectable/global_effectable.cpp [hunk 14]
```diff
@@ -334,11 +354,13 @@ bool GlobalEffectable::modEncoderButtonAction(uint8_t whichModEncoder, bool on,
 	else if (modKnobMode == 1) {
 		if (whichModEncoder == 1) {
 			if (on) {
-				currentFilterType =
-				    static_cast<FilterType>((util::to_underlying(currentFilterType) + 1) % kNumFilterTypes);
-
+				// Get the current value to a tmp variable
+				FilterType tmpFilterType = currentFilterType;
+				// Change the tmp value
+				tmpFilterType = static_cast<FilterType>((util::to_underlying(tmpFilterType) + 1) % kNumFilterTypes);
```
Comment (seangoodvibes) [inline src/deluge/model/global_effectable/global_effectable.cpp:360]: leftover

**Review by seangoodvibes — CHANGES_REQUESTED** [review]

**Pull request closed without merging**